image-definitions/README.md

```diff
@@ -112,14 +115,27 @@ In many cases only single image should be published. In such cases perform publi
 
 ### Maintenance
 
-From time to time an update in upstream source, an addition of new container image, or a new vulnerability fix will require re-evaluation of image definitions. To perform difference analysis between upstream, set up a clean copy of upstream source in temporary directory and diff the contents with current image definitions.
+To create initial patch setup a clean copy of upstream source, make required adjustments that would resolve CVEs and produced the initial patch:
 
 For Kubeflow:
 
 ```
-mkdir ./update
-setup.sh ./update
-diff -r ./update/kubeflow kubeflow
+setup.sh .
+cd kubeflow
+git diff > ../kubeflow.patch
+```
+
+Commit this patch to this repository.
+
+From time to time an update in upstream source, an addition of new container image, or a new vulnerability fix will require re-evaluation of image definitions. To perform difference analysis between upstream, set up a clean copy of upstream source, apply existing patch and diff the contents with current image definitions.
+
+For Kubeflow:
+
+```
+setup.sh .
+cd kubeflow
+git apply ../kubeflow.patch
+git diff
```

Comment:
I might be getting this wrong, but I can't get this workflow to work if there's a change from upstream.

Consider this:
* the current committed patch of `kubeflow.patch` changes the `components/crud-web-apps/jupyter/Dockerfile`'s first line from using `python:3.7-slim-buster` to `python:3.7-slim-bullseye` (see below for the patch contents)
* say upstream modifies their version of `components/crud-web-apps/jupyter/Dockerfile`, so instead of using `python:3.7-slim-buster` they now use `python:3.7-slim-shady`.  To pretend, we can manually change the files in `./kubeflow` to reflect this
* `git apply ../kubeflow.patch` fails because the patch file is not valid.  The patch file tries to change `slim-buster` to `slim-bullseye`, but it fails without doing anything because it can't find the relevant place to patch

Am I doing something wrong on this one?

patch details for reference:
```
diff --git a/components/crud-web-apps/jupyter/Dockerfile b/components/crud-web-apps/jupyter/Dockerfile
index d68af3fe..fe623656 100644
--- a/components/crud-web-apps/jupyter/Dockerfile
+++ b/components/crud-web-apps/jupyter/Dockerfile
@@ -1,5 +1,5 @@
 # --- Build the backend kubeflow-wheel ---
-FROM python:3.7-slim-buster AS backend-kubeflow-wheel
+FROM python:3.7-slim-bullseye AS backend-kubeflow-wheel
```

Reply:
So, this is the problem that we have to live with. In such situations we need to analyze what the issue is and not blindly apply the patch. The error `git apply` gives is perfect indicator of something like this happening.
Your example is what will actually happen with our patch tacking when they start fixing vulnerabilities in upstream. What if they updated base image to `python:3.8-ubuntu`? Then we would need to manually modify files and create a new patch. If we like what they change we remove those lines from our patch. If we still don't like their change we will manually modify `Dockerfile` and create new patch.

Reply:
ok if that's what is expected, that makes sense.  Could you add a small example of how someone should deal with that in the readme?  When I tried it and started getting `patch failed: ...` messages, I figured I had done something wrong

Reply:
Done.